### ISSUE TITLE
Incremental parser leaks space?

### ISSUE THREAD

**Issue by ethercrow** [yi-editor/yi#105]
_Original author: reiner.p...@gmail.com (May 24, 2011 08:45:03)_

<b>What steps will reproduce the problem?</b>
Use yi for a while, editing Haskell files with the Haskell mode (cleverMode). Over time, the memory consumption increases. I have found it quite typical for yi to consume around 300MB after I have used it for around 30mins.

Running 'reload' drops the memory use to much less -- often around 30MB.

The attached heap profile suggests that the incremental parser is mostly responsible.

_Original issue: http://code.google.com/p/yi-editor/issues/detail?id=356_

**Comment by siddhanathan** [comment]
The issue seems to be that the incremental parser uses too much memory (confirmed in Haskell precise mode as well). The more the file gets traversed, the higher the memory consumption. The memory consumption does seem to stay constant once the entire file has been traversed, so the issue is just that the parser uses too much space. This is a big issue, since, the amount of space required to traverse the entire file is about 2GB for a Haskell file which is 50k sloc and 1MB in size.

**Comment by siddhanathan** [comment]
Here's a heap profile using Haskell mode to show the space leak in the Incremental parser:

![heap_profile](https://cloud.githubusercontent.com/assets/933964/21085979/0e197c32-bfcd-11e6-904a-29806e451f5b.png)

Syntax highlighting for Python does much better in this regard:

![better_heap_profile](https://cloud.githubusercontent.com/assets/933964/21085997/2d85f78a-bfcd-11e6-922e-c327e9f32c12.png)

The operations I was performing was traversing the file with no changes to the file itself. Towards the end, you can see that the memory usage stays consistent since the entire file has been traversed.

I can verify that the parser runs in constant memory with the default mode (that is, no syntax highlighting). This is what the heap profile look like in this case:

![heap_profile_default](https://cloud.githubusercontent.com/assets/933964/21086164/b75f3b5a-bfce-11e6-8545-ce4b4ad002f8.png)

So Haskell mode is forcing the Incremental parser to evaluate unnecessarily.

**Comment by siddhanathan** [comment]
Profiling data seems to indicate that our culprit is getIndexedStream in https://github.com/yi-editor/yi/blob/c2d8d7eeb2567ee4634a523b68bef08cd0747ce7/yi-core/src/Yi/Buffer/Implementation.hs#L431 which is forcing the text. We would ideally like to be able to traverse the YiString somehow without keeping it in memory.

**Comment by noughtmare** [comment]
Maybe the `Scanner`s can be replaced by something like `pipes` or `machines` or `conduit`s?

**Comment by noughtmare** [comment]
From the comment it seems that @Fuuzetsu already suspected `getIndexedStream` would become a bottleneck.
> -- | TODO: This guy is a pretty big bottleneck and only one function
> -- uses it which in turn is only seldom used and most of the output is
> -- thrown away anyway. We could probably get away with never
> -- converting this to String here. The old implementation did so
> -- because it worked over ByteString but we don't have to.

Also

> So Haskell mode is forcing the Incremental parser to evaluate unnecessarily.

I think it is not the bug in the `haskell-mode`, but rather a general bug in all syntax highlighters.

**Comment by siddhanathan** [comment]
@noughtmare Yes. There was a discussion about this a long time ago on IRC. Should have had the discussion here instead. The issue is not specific to Haskell mode, rather, it's syntax highlighting in general.

I think getIndexedStream was suspected to be a CPU bottleneck. Here it's causing a space leak. Not exactly the same, but getting rid of it is beneficial regardless.

Modifying the syntax highlighter such that it doesn't need to use getIndexedStream would probably be the way to go: https://github.com/yi-editor/yi/blob/51855c9971c1e611b2c823de2ac4bd068f6dfb78/yi-language/src/Yi/Syntax.hs#L110

Or, at the very least, figure out why it's forcing the incremental parser.

**Comment by Fuuzetsu** [comment]
>Modifying the syntax highlighter such that it doesn't need to use getIndexedStream would probably be the way to go

Yes, the problem is that IIRC it's non-trivial which is why the function stuck around to begin with.

**Comment by noughtmare** [comment]
Why is the getIndexedStream function bad, is it because it returns a list? Could we use one of the streaming libraries (like pipes or conduits) to circumvent this?

**Comment by Fuuzetsu** [comment]
Everything is implemented in terms of `YiString` (which IIRC provided great speedup and much joy all around). It's implemented as a rope with chunks of `Text`. Here we call toString which unchunks the tree, unpacks concats the Text chunks, unpacks into String, potentially reverses the whole thing (yes, whole buffer), throws away some part of it and then returns the rest of the string chunked into 1 character each. As you can imagine, this isn't fast. Streaming libraries don't really help because implementation will have to be changed anyway.

I suppose the least painful improvement would be to write some functions in `yi-rope` which instead of doing all of the above provided the caller with some sort of an iterator that walks over the YiString directly. This could be better.

I have a bunch of time off work from tomorrow so maybe I could look into it if you can't/don't want to but I've been known to say that then disappear!

**Comment by noughtmare** [comment]
Would just using a foldl (or foldl') be more memory efficient than the current situation?
We could then make a field of the Highlighter be a function like:
```haskell 
hlStep :: State a -> Text -> State a
```
Or do you mean that 'walking' the YiString from 'left' to 'right' is not the way to go?

**Comment by siddhanathan** [comment]
@Fuuzetsu go for it. I'll try picking up in a few weeks if you do disappear.

**Comment by noughtmare** [comment]
> I could look into it if you can't/don't want to

I intend to look into this further (I think this issue might be related to the problems I'm facing with #946). But any help is very much appreciated!

**Comment by Fuuzetsu** [comment]
I'm saying we can walk it but the way we're doing it right now is awful. It might have worked many years ago when the stream would have been produced lazily but in this case I believe we're going all out and then walking a forced stream. I think, I don't have data on hand.

Profile the function and see what it does.

**Comment by noughtmare** [comment]
The fastMode for haskell seems to be the exact same as the clevermode but then without the Incremental Parser overhead.

fastMode:
![yi](https://cloud.githubusercontent.com/assets/2665074/21471059/769745ba-caa3-11e6-9ede-30782c095fd6.png)

cleverMode:
![yi](https://cloud.githubusercontent.com/assets/2665074/21471054/38226292-caa3-11e6-9146-f14fe5df1551.png)

**Comment by noughtmare** [comment]
Ok, I think I found the problem. The Highlighter caches every step of the making of the abstract syntax tree All these states are being cached to speed up the generation of the AST when the buffer is edited (which naturally takes a lot of memory). 

A solution to this would be to not cache any state at all, at the cost of more cpu time. 

I think that we could also save a lot of memory by caching the state between a point and the previous point instead of caching all state up to a point for every point (this might not help because of laziness).

PS: I don't think this bug is caused by the transition to the yi-rope type, because it was filed in 2011 (I believe yi-rope was made around 2014).

**Comment by siddhanathan** [comment]
@noughtmare Can you verify that? I believe I tried turning caching off at some point, and didn't see a difference.

**Comment by noughtmare** [comment]
I'm working on it, I could just be confused by laziness.

**Comment by noughtmare** [comment]
I have some results:
The `unfoldLexer` function was to blame for the largest memory usage (in the fastMode for haskell). Just chaning it to:
```haskell
unfoldLexer :: ((state, input) -> Maybe (token, (state, input)))
            -> (state, input) -> [(state, token)]
unfoldLexer f b = fst b `seq` snd b `seq` case f b of
             Nothing -> []
             Just (t, b') -> (fst b, t) : unfoldLexer f b'
```
Removes that memory leak.

**Comment by noughtmare** [comment]
Here are the graphs.

fastMode:
![yi](https://cloud.githubusercontent.com/assets/2665074/21502393/bb455276-cc4f-11e6-8ca2-c701f5deebe8.png)

cleverMode:
![yi](https://cloud.githubusercontent.com/assets/2665074/21502356/8720dbc8-cc4f-11e6-8643-5cf573e4de27.png)

**Comment by Fuuzetsu** [comment]
What's your test.hs?

**Comment by noughtmare** [comment]
same as https://github.com/yi-editor/yi/issues/105#issuecomment-269120677.
A 618 KB haskell file with just the Data.Data module (first thing I found) pasted into it many times.

**Comment by siddhanathan** [comment]
@Fuuzetsu This pattern should work:

~~~hs

main :: IO ()
main = do
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
    putStrLn "Hèllo"
~~~

going on till infinity.... (95k lines in my case)

**Comment by Fuuzetsu** [comment]
OK, I'm going to have some fun I think; do you have any more changes except the above strictness change?

**Comment by noughtmare** [comment]
No, that's it.

**Comment by noughtmare** [comment]
I get even better space usage when adding strictness in Yi.IncrementalParse.scanner.updateState0:
```haskell
updateState0 curState toks@((st,tok):rest) = nextState `seq` ((st, curState), result) : updateState0 nextState rest
```
![yi](https://cloud.githubusercontent.com/assets/2665074/21509879/ed33926e-cc8d-11e6-91b5-a3fbf65cfcc4.png)

**Comment by Fuuzetsu** [comment]
Both the strictness annotations (use `BangPatterns` instead of `seq`) while potentially useful (in unfoldLexer the tuple constructor can be forced without bad effect at least I believe) miss the main issue. You should ask yourself "why am I seeing 200-300MB when I open the file"?

It can be quickly answered by looking into the incremental parsing paper (see README) then starting Yi on big file, editing first line then checking the heap. You'll find that memory usage does not go down when you compare this to going to the end of the file, back and editing the first line. Notably, we're *not* parsing incrementally currently. The whole point of the incremental parser is that we only parse up to the point we display on screen. If we change something then everything past the point of change to the point of current view should be parsed but nothing more. I recommend retainer profiling with `-R20` if you want to see what needs fixing for yourself. I have found what I think is the main issue and checking potential fixes.

**Comment by Fuuzetsu** [comment]
I suppose a picture speaks thousand words so

![1482886543](https://cloud.githubusercontent.com/assets/893115/21511405/73c39680-cc98-11e6-8e49-02b050c61f72.png)

This is loading nearly 1MB file with clever Haskell, down from 300-350MB to ~5MB: just opening file without moving anywhere. I still see some other problems though so not submitting anything yet.

There is of course a lot of value of looking for improving worst case (i.e. what you're seeing above) so I encourage you to pursue it but just be aware of what you're measuring.

**Comment by siddhanathan** [comment]
Interesting. Wasn't aware of retainer profiling. Makes it so much easier.

Fixing the memory issue in Vty refresh function [(Vty.hs#L212)](https://github.com/yi-editor/yi/blob/51855c9971c1e611b2c823de2ac4bd068f6dfb78/yi-frontend-vty/src/Yi/Frontend/Vty.hs#L212) with that becomes fairly trivial:

~~~hs
        (miniImages, bigImages) = let f = fmap (picture . snd) in bimap f f
                                $ partition (isMini . fst) (toList windowsAndImages)
~~~

@Fuuzetsu are you considering Pango as well?

**Comment by Fuuzetsu** [comment]
@siddhanathan There is a much worse problem than lack of sharing there. The sizes used for rendering text and obtaining strokes is completely wrong: it's using Rect size which is a pixel size to specify number of characters; so on 1080p screen, we're saying "render next 2million characters". This was responsible for whole file being parsed as soon as you opened. The fix there is simple (and can be improved) and gets rid of indexed stream usage. I inline some of it here though I'll likely send a PR some time.

```
@@ -66,6 +66,7 @@ import           Yi.Config
 import           Yi.Debug                       (logError, logPutStrLn)
 import           Yi.Editor
 import           Yi.Event                       (Event)
+import qualified Yi.Rope as R
 import           Yi.Style
 import           Yi.Types                       (YiConfigVariable)
 import qualified Yi.UI.Common                   as Common
@@ -74,7 +75,7 @@ import           Yi.Layout                      (HasNeighborWest)
 import           Yi.UI.TabBar                   (TabDescr (TabDescr), tabBarDescr)
 import           Yi.UI.Utils                    (arrangeItems, attributesPictureAndSelB)
 import           Yi.Frontend.Vty.Conversions          (colorToAttr, fromVtyEvent)
-import           Yi.Window                      (Window (bufkey, isMini, wkey))
+import           Yi.Window                      (Window (bufkey, isMini, wkey, width, height))
 
 
 data Rendered = Rendered
@@ -240,10 +241,12 @@ refresh fs e = do
         { Vty.picCursor = cursorPos }
 
 renderWindow :: UIConfig -> Editor -> SL.Rect -> HasNeighborWest -> (Window, Bool) -> Rendered
-renderWindow cfg e (SL.Rect x y w h) nb (win, focused) =
+renderWindow cfg e (SL.Rect x y _ _) nb (win, focused) =
     Rendered (Vty.translate x y $ if nb then vertSep Vty.<|> pict else pict)
              (fmap (\(i, j) -> (i + y, j + x')) cur)
     where
+        w = Yi.Window.width win
+        h = Yi.Window.height win
         x' = x + if nb then 1 else 0
         w' = w - if nb then 1 else 0
         b = findBufferWith (bufkey win) e
@@ -257,21 +260,25 @@ renderWindow cfg e (SL.Rect x y w h) nb (win, focused) =
         wsty = attributesToAttr ground Vty.defAttr
         eofsty = appEndo (eofStyle sty) ground
         (point, _) = runBuffer win b pointB
-        region = mkSizeRegion fromMarkPoint $ Size (w' * h')
+        (text, _) = runBuffer win b $
+          -- Take the window worth of lines; we now know exactly how
+          -- much text to render, parse and stroke.
+          fst . R.splitAtLine h' <$> streamB Forward fromMarkPoint
+
+        region = mkSizeRegion fromMarkPoint . Size $! R.length text
         -- Work around a problem with the mini window never displaying it's contents due to a
         -- fromMark that is always equal to the end of the buffer contents.
         (Just (MarkSet fromM _ _), _) = runBuffer win b (getMarks win)
         fromMarkPoint = if notMini
                         then fst $ runBuffer win b $ use $ markPointA fromM
                         else Point 0
-        (text, _) = runBuffer win b (indexedStreamB Forward fromMarkPoint)
 
         (attributes, _) = runBuffer win b $ attributesPictureAndSelB sty (currentRegex e) region
         -- TODO: I suspect that this costs quite a lot of CPU in the "dry run" which determines the window size;
         -- In that case, since attributes are also useless there, it might help to replace the call by a dummy value.
         -- This is also approximately valid of the call to "indexedAnnotatedStreamB".
         colors = map (fmap (($ Vty.defAttr) . attributesToAttr)) attributes
-        bufData =  paintChars Vty.defAttr colors text
+        bufData =  paintChars Vty.defAttr colors $! zip [fromMarkPoint..] (R.toString text)
```

However it doesn't fix the problem fully: if we go back to the start of file and change it, the memory does not go back to what it was when we initially opened the file. Something is holding onto parser state and that's what I'm trying to find now.

Lastly, no, I'm not looking at Pango right this second; I want to finally remove all the gtk2 code and migrate it to gtk3 before doing anything else there. It also obfuscates profiling information because it's much more resource intensive than Vty so I'm working with that right now.

**Comment by noughtmare** [comment]
I get this strange graph now:
![yi](https://cloud.githubusercontent.com/assets/2665074/21521369/8c101678-ccfc-11e6-81bb-eb3140a38fd5.png)
I opened the test.hs file and then pressed G and waited a few seconds and then started scrolling up for a few seconds.

**Comment by siddhanathan** [comment]
The profiler seems to be pointing towards updateCache in the highlighter: https://github.com/yi-editor/yi/blob/51855c9971c1e611b2c823de2ac4bd068f6dfb78/yi-language/src/Yi/Syntax.hs#L117

Becomes clear when we add `-L100` to the RTS options.

**Comment by siddhanathan** [comment]
Reached the same conclusion @Fuuzetsu reached a while ago. Something is holding on to the Parser state, but not using it.

![bio](https://cloud.githubusercontent.com/assets/933964/21555876/1d732f68-cdea-11e6-95c3-365ba4eb58a6.png)

![retainers](https://cloud.githubusercontent.com/assets/933964/21555906/6b234d60-cdea-11e6-8237-e6c66d8f831c.png)

**Comment by siddhanathan** [comment]
@noughtmare Instead of forcing values in unfoldLexer, we could make them lazier.

```diff
diff --git a/yi-language/src/Yi/Lexer/Alex.hs b/yi-language/src/Yi/Lexer/Alex.hs
index ccd15471..f1ba7be6 100644
--- a/yi-language/src/Yi/Lexer/Alex.hs
+++ b/yi-language/src/Yi/Lexer/Alex.hs
@@ -212,9 +212,9 @@ lexScanner Lexer {..} src = Scanner
 -- | unfold lexer into a function that returns a stream of (state, token)
 unfoldLexer :: ((state, input) -> Maybe (token, (state, input)))
             -> (state, input) -> [(state, token)]
-unfoldLexer f b = case f b of
+unfoldLexer f ~(s, i) = case f (s, i) of
              Nothing -> []
-             Just (t, b') -> (fst b, t) : unfoldLexer f b'
+             Just ~(t, ~b') -> (s, t) : unfoldLexer f b'
 
 -- * Lenses
 makeLensesWithSuffix "A" ''Posn
```

This has the same effect, and reduces memory consumption in fastmode by over 50% in my tests. Using strictness annotations here would unnecessarily force the incremental parser.

**Comment by siddhanathan** [comment]
@noughtmare I don't think that does anything. I just put it there for emphasis. I know `b'` is a tuple and the function call will evaluate it lazily.

**Comment by Fuuzetsu** [comment]
The strictness change that helps there is not due to unforced values at all, it's the tuple constructor that's not being matched on and we just build up thunks. Just matching on that our using bang on 'b' should have same effect.

**Comment by siddhanathan** [comment]
There is one small change I have that reduces memory consumption further. It changes the syntax driver to use non-incremental parsing (which is not an issue, since syntax highlighting happens in one shot anyway).

**Comment by Fuuzetsu** [comment]
It is an issue because it relies on a bug.  Highlighting requests strokes for specific region and that region is the visible area of window (with patch I posted). Gaining memory improvement by assuming full buffer parsing is wrong

**Comment by noughtmare** [comment]
Also interesting is what happens when you just open the file, wait some time and then close it again:
![yi-arr-words](https://cloud.githubusercontent.com/assets/2665074/21642037/f1a114aa-d27f-11e6-8aa8-a0898d012679.png)
This is using the fastMode.

As you can see. The ARR_WORDS (this is the internal constructor of the ByteString and Text datatypes) data constructor takes up more space than I expected. 

Is it normal for a YiString to take up ~18x the amount of memory of the raw file?

I also tested this with no mode enabled and it gave a very similar picture (around 11MB total usage).

**Comment by noughtmare** [comment]
I think I've succeeded in forcing the YiString to be kept in memory, here's the heap:
![rope-test](https://cloud.githubusercontent.com/assets/2665074/21644162/7c37a528-d28c-11e6-82d8-43c24f4f8466.png)
Code:
```haskell
module Main where

import Prelude hiding (readFile, putStrLn)
import Yi.Rope (toText, readFile)
import Data.Text.IO (putStrLn)

main :: IO ()
main = do
  rope <- readFile "test.hs"
  case rope of
    Right (r,_) -> do
      putStrLn (toText r)
      test 100000000
      putStrLn (toText r)
    Left _ -> error "left"
  where test n | n == 0 = return () | otherwise = test (n - 1)
```
From this we can see that the rope uses around 2x more memory.

**Comment by siddhanathan** [comment]
IIRC the YiString is fully evaluated when opening the file, so that memory usage shouldn't go any higher. 18x doesn't sound too bad to me.

**Comment by noughtmare** [comment]
@siddhanathan As you can see from [my previous comment](https://github.com/yi-editor/yi/issues/105#issuecomment-270373925), the memory usage of YiString is only 2.5x so I think 18x can be improved.

**Comment by siddhanathan** [comment]
For what it's worth, here's a change that eliminates caching:

~~~diff
diff --git a/yi-core/src/Yi/Syntax/Driver.hs b/yi-core/src/Yi/Syntax/Driver.hs
index f62a4b7e..ca540423 100644
--- a/yi-core/src/Yi/Syntax/Driver.hs
+++ b/yi-core/src/Yi/Syntax/Driver.hs
@@ -31,19 +31,12 @@ mkHighlighter scanner =
         , hlGetTree      = \(Cache _ _ _ focused) w -> M.findWithDefault emptyResult w focused
         , hlFocus        = focus
         }
-    where startState :: state
-          startState = scanInit    (scanner emptyFileScan)
+    where 
           emptyResult = scanEmpty (scanner emptyFileScan)
           updateCache :: Scanner Point Char -> Point -> Cache state tree tt -> Cache state tree tt
-          updateCache newFileScan dirtyOffset (Cache path cachedStates oldResult _) = Cache path newCachedStates newResult M.empty
+          updateCache newFileScan dirtyOffset (Cache path _states oldResult _) = Cache path _states newResult M.empty
             where newScan = scanner newFileScan
-                  reused :: [state]
-                  reused = takeWhile ((< dirtyOffset) . scanLooked (scanner newFileScan)) cachedStates
-                  resumeState :: state
-                  resumeState = if null reused then startState else last reused
-
-                  newCachedStates = reused ++ fmap fst recomputed
-                  recomputed = scanRun newScan resumeState
+                  recomputed = scanRun newScan (scanInit newScan)
                   newResult :: tree (Tok tt)
                   newResult = if null recomputed then oldResult else snd $ head recomputed
           focus r (Cache path states root _focused) =
~~~

This fixes the memory problem.

**Comment by noughtmare** [comment]
Ah, now I know why I didn't see any memory improvements when I tried to disable caching. We have two `mkHighlighter` functions one in `Yi.Syntax.Driver` and one in `Yi.Syntax`. Only the javascript mode uses that other `mkHighlighter`, is there any reason to have two of these functions?

**Comment by noughtmare** [comment]
I made some graphs with just the change @siddhanathan proposed to disable caching. I didn't use any other modifications. To generate these I first opened the file moved to the bottom and then moved the cursor up and deleted one character then moved the cursor around some more and reinserted the character and then moved around some more and then closed the editor. 

Precise mode:
![yi-no-cache](https://cloud.githubusercontent.com/assets/2665074/21719611/ed3442c0-d41e-11e6-9015-ab43397ce73e.png)

cleverMode:
![yi-no-cache-clever](https://cloud.githubusercontent.com/assets/2665074/21719794/e15b7ab2-d41f-11e6-9f84-f8d36d1d4c2b.png)

fastMode:
![yi-no-cache-fast](https://cloud.githubusercontent.com/assets/2665074/21719802/e5f985a0-d41f-11e6-8be7-1574d148eafe.png)

**Comment by noughtmare** [comment]
The retention profile of the cleverMode:
![yi-no-cache-clever-retention](https://cloud.githubusercontent.com/assets/2665074/21720518/f4ce5fb6-d423-11e6-82b2-53010cd60949.png)

**Comment by noughtmare** [comment]
Making apply strict:
```diff
diff --git a/yi-core/src/Parser/Incremental.hs b/yi-core/src/Parser/Incremental.hs
index 01961935..b97881a6 100644
--- a/yi-core/src/Parser/Incremental.hs
+++ b/yi-core/src/Parser/Incremental.hs
@@ -149,7 +149,7 @@ instance Show (RPolish i o) where
 
 
 apply :: forall t t1 a. ((t -> a) :< (t :< t1)) -> a :< t1
-apply ~(f:< ~(a:<r)) = f a :< r
+apply (f:<(a:<r)) = f a `seq` f a :< r
```
Results in (cleverMode):
![yi-no-cache-clever-retention-apply](https://cloud.githubusercontent.com/assets/2665074/21721184/7c68213e-d427-11e6-91ac-9278eb45cbb8.png)

This makes it seem like `flushAccumulatorE` is the problem.

**Comment by siddhanathan** [comment]
I'm curious about the performance impact of the patch I posted above. And maybe evaluating other ways of having a cache that doesn't keep the incremental parser state around unnecessarily. That patch simply makes syntax highlighting non-incremental, which might not be what we want.

> Making apply strict

Experimenting is fine. But there's a very good reason that application is lazy. Rather than blaming the incremental parser, try blaming the entity using the incremental parser incorrectly. In this case, what is doing the call to apply, and is that call retaining memory unnecessarily (checked via a biography profile).

**Comment by noughtmare** [comment]
```
yi: internal error: Invalid object in processHeapClosureForDead(): 131184
    (GHC version 8.0.1 for x86_64_unknown_linux)
    Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
[1]    11407 abort (core dumped)  yi -fvty -kvim test.hs +RTS -hbdrag,void -hc
```
😞

**Comment by siddhanathan** [comment]
@noughtmare Leave a comment here https://ghc.haskell.org/trac/ghc/ticket/7836 . If you have reproducible steps, then mention them in the bug.

**Comment by siddhanathan** [comment]
I'm guessing some part of syntax highlighting is holding on to the lexer state.


Clever mode:

![alex](https://cloud.githubusercontent.com/assets/933964/21738839/8a832f9c-d441-11e6-9da7-1749ce5c125b.png)

Fast mode:

![fastmode](https://cloud.githubusercontent.com/assets/933964/21739045/0f56e878-d445-11e6-82f5-5a98822b370c.png)

**Comment by noughtmare** [comment]
More results!

```diff
diff --git a/yi-keymap-vim/src/Yi/Keymap/Vim/StateUtils.hs b/yi-keymap-vim/src/Yi/Keymap/Vim/StateUtils.hs
index a694c1f6..b288e2c1 100644
--- a/yi-keymap-vim/src/Yi/Keymap/Vim/StateUtils.hs
+++ b/yi-keymap-vim/src/Yi/Keymap/Vim/StateUtils.hs
@@ -91,7 +91,7 @@ flushAccumulatorE :: EditorM ()
 flushAccumulatorE = do
     accum <- vsAccumulator <$> getEditorDyn
     let repeatableAction = stringToRepeatableAction accum
-    modifyStateE $ \s ->
+    accum `seq` repeatableAction `seq` modifyStateE $ \s ->
         s { vsRepeatableAction = Just repeatableAction
           , vsAccumulator = mempty
           , vsCurrentMacroRecording = fmap (fmap (<> accum))
diff --git a/yi-language/src/Yi/Lexer/Alex.hs b/yi-language/src/Yi/Lexer/Alex.hs
index ccd15471..a9a273e7 100644
--- a/yi-language/src/Yi/Lexer/Alex.hs
+++ b/yi-language/src/Yi/Lexer/Alex.hs
@@ -212,7 +212,7 @@ lexScanner Lexer {..} src = Scanner
 -- | unfold lexer into a function that returns a stream of (state, token)
 unfoldLexer :: ((state, input) -> Maybe (token, (state, input)))
             -> (state, input) -> [(state, token)]
-unfoldLexer f b = case f b of
+unfoldLexer f b = fst b `seq` case f b of
              Nothing -> []
              Just (t, b') -> (fst b, t) : unfoldLexer f b'
```

before:
![yi-before](https://cloud.githubusercontent.com/assets/2665074/22073475/b7a71dfe-dda5-11e6-9fe6-bd3f10414d01.png)

after:
![yi-after](https://cloud.githubusercontent.com/assets/2665074/22073488/bd75f534-dda5-11e6-8695-55de6bbb325f.png)

I've gotten rid of the memory leak 😃!

**Comment by noughtmare** [comment]
One more thing.
My routine for this test was first open a big file (same as above), then press 'G' (to force the highlighter) then 'vggd' (delete the entire file) then 'i' and entering some random text, then 'upG' (undoing, pasting in the original file and scrolling down) and then some cursor movement and finally ':q!' (to quit).

As you can see in the after picture there is a moment (around the 2 sec mark) where the memory is constant. I think this is the time when I entered the random text, I think it should be possible to already release the memory there, but I don't know how.

**Comment by noughtmare** [comment]
I forgot to add results for other modes.

Precise mode after:
![yi-after-precise](https://cloud.githubusercontent.com/assets/2665074/22078121/423b5b9a-ddb7-11e6-98be-7126e1c8e41d.png)

**Comment by noughtmare** [comment]
I forgot to add results for other modes.

Precise mode:
![yi-after-precise](https://cloud.githubusercontent.com/assets/2665074/22078386/14a0aeaa-ddb8-11e6-8810-3bef25289897.png)

**Comment by siddhanathan** [comment]
So if I understand correctly, it makes things worse in precise mode. As Fuuzetsu mentioned, a bang pattern on `b` does the job perfectly fine, and should give similar results in both modes.

What is actually happening is quite simple. The syntax driver holds on to the list of parser states in its cache. This is okay, because it's not using those parser states before the user starts editing the text. To verify that, ensure that `updateCache` is called only once when casually scrolling. The space complexity of this is `O(n)` as proposed by Jean-Philippe Bernardy, where n is the cursor position where the edit occurs. When no edits occur, this should use negligible space. On the contrary, what we're seeing is that scrolling the window pane forces this list. My guess would be that some part of the program related to focusing is evaluating the parser state independently and throwing it away. Even though that seems harmless, because of sharing and memoization, it has the side effect of forcing the list of parser states held in the syntax driver.

The other issue that I had since the very beginning is, even if we do manage `O(n)` space complexity, the actual space being used is a lot. The way incremental parsing is done right now essentially rules out the possibility of using Yi with big data files. It is a tradeoff between space/time, and that's understandable, but it's a tradeoff we need to be aware about (gigabytes of memory for a megabyte text file is a lot!). One way to reduce the space complexity further is to parse chunks at a time (and this might fit nicely with the rope structure we are using), and store the state of parsing chunks. This way, we don't store the intermediate parse state after parsing every character. The results would be less online-ish, but at least use significantly less space.

@ethercrow @Fuuzetsu any thoughts on replacing the character scanners with chunk scanners?

**Comment by noughtmare** [comment]
The performance has not degraded, these I will post the befores.

Precise:
![yi-before-precise](https://cloud.githubusercontent.com/assets/2665074/22081455/f2487fd4-ddc3-11e6-9314-3ec9726f76ef.png)

**Comment by noughtmare** [comment]
This is strange, I just press 'G' in the file and then right after that ':q!':
![yi-strange](https://cloud.githubusercontent.com/assets/2665074/22086432/fb434492-ddd8-11e6-964d-355b7a02f9ac.png)

**Comment by noughtmare** [comment]
I just noticed that the precise parser is (even without the strictness changes) not usable. I opened yi-core/src/Parser/Incremental.hs and tried to pressed x somewhere in the middle of the file and Yi just froze and started consuming all my memory. I had to kill it forcefully.

**Comment by noughtmare** [comment]
I dicovered something else. The large spikes in the precise mode are due to a bug in the parser, here's a heap of another file with ~20K lines (657KB):

![yi-withoutbug](https://cloud.githubusercontent.com/assets/2665074/22349058/4e60a67a-e40e-11e6-9f69-e9f612f7bfc2.png)

350 MB is a lot better than 1GB.

**Comment by noughtmare** [comment]
The new picture of the scroll down -> delete entire buffer -> waste some time -> paste -> scroll down again, looks very much like the clevermode one:

![yi-without](https://cloud.githubusercontent.com/assets/2665074/22349841/bf6c74cc-e411-11e6-8c55-fa15cbf7f981.png)

**Comment by noughtmare** [comment]
The strange thing about the incrementalparse parsers is that the toP and iBest data persists even when the entire buffer is deleted and it doesn't generate a new data when pasting it back in, which makes me believe the memory usage is unnecessary.

**Comment by noughtmare** [comment]
More news:
[The nocache syntax driver proposed]( https://github.com/yi-editor/yi/issues/105#issuecomment-270837230) by @siddhanathan makes the preciseMode use only 25MB! This gives me even more confidence in the fact that the 300MB memory usage is unnecessary!

**Comment by noughtmare** [comment]
I think I fixed it:

```diff

diff --git a/yi-core/src/Yi/Syntax/Driver.hs b/yi-core/src/Yi/Syntax/Driver.hs
index f62a4b7e..ca0e5ca1 100644
--- a/yi-core/src/Yi/Syntax/Driver.hs
+++ b/yi-core/src/Yi/Syntax/Driver.hs
@@ -45,7 +45,7 @@ mkHighlighter scanner =
                   newCachedStates = reused ++ fmap fst recomputed
                   recomputed = scanRun newScan resumeState
                   newResult :: tree (Tok tt)
-                  newResult = if null recomputed then oldResult else snd $ head recomputed
+                  newResult = if null recomputed then oldResult else snd $ head $ scanRun newScan $ if null reused then scanInit (scanner newFileScan) else last reused
           focus r (Cache path states root _focused) =
               Cache path' states root focused
               where (path', focused) = unzipFM $ zipWithFM (\newpath oldpath -> fromNodeToFinal newpath (oldpath,root)) [] r path

```

Precise mode:

![yi-fixed](https://cloud.githubusercontent.com/assets/2665074/22383670/0268d57c-e4cb-11e6-9c9e-d962c8986c92.png)

**Comment by noughtmare** [comment]
Unfortunantly, inserting text at the end of the file still causes 300MB to be used.

**Comment by siddhanathan** [comment]
> Unfortunantly, inserting text at the end of the file still causes 300MB to be used.

Wait, that is the behavior we want. Scrolling through the file shouldn't cause any memory consumption.

**Comment by noughtmare** [comment]
@siddhanathan Shouldn't we want to save intermediate states while parsing and parse while scrolling?

I think that the reason for the high memory usage is that we're reparsing the file for every intermediate state that we cache.

The `mkHighlighter` function only uses the head of the recomputed list of states, which is essential for the lazy parsing, but it also means that its not caching the intermediate states while parsing.

**Comment by siddhanathan** [comment]
> Shouldn't we want to cache intermediate states while parsing and parse while scrolling (and therefore consume memory while scrolling)?

Parsing is an expensive operation. So no. It would cause scrolling to be slow.

> it at least parses the file twice, once to get the final result and another time to find the intermediate state (which defeats the purpose of caching the states).

using different representations, so this should hopefully be okay.

> I think that the reason for the high memory usage is that we're reparsing the file for every intermediate state that we cache.

Maybe. I'm not sure. I can verify that your patch does improve memory usage.

**Comment by noughtmare** [comment]
Ah, I meant that we're parsing while scrolling down for the first time.

**Comment by siddhanathan** [comment]
> Ah, I meant that we're parsing while scrolling down for the first time

Yes. That bothers me though. I don't think that should happen. My guess is, something somewhere is innocently using the parser without requiring too much memory - without realizing that it is forcing the parse state which results in higher memory usage for the syntax driver. Which is also why identifying the problem is difficult, it's not the function with the highest memory usage which is the root cause here.

Now we do need high memory usage when editing the file, but that is proportional to the position of the edit, so if the user edits the file at the beginning, I would hope the memory usage would not be very high. Lowering this memory consumption would be a separate issue, because that could compromise performance.

**Comment by noughtmare** [comment]
I think it was designed to lazily parse the file so that it can display the file immediately on opening.

from [yi-editor.github.io](http://yi-editor.github.io/posts/2014-09-04-incremental-parsing/):

> Hughes and Swierstra have shown how online parsers can be constructed. An online parser takes advantage of lazy evaluation: the input is analyzed **only as far as needed** to produce the part of the result that is forced.

**Comment by noughtmare** [comment]
> using different representations, so this should hopefully be okay.

It should work, but only after the first change. The first change would then cause the state to be calculated, it will only use the saved state on subsequent changes.

**Comment by Fuuzetsu** [comment]
>Yes. That bothers me though. I don't think that should happen.

It should and it's by design: how else are we going to display syntax highlighting to the user if we don't parse until the point user is viewing?
